pkg/handler/collector/s3/s3.go

```diff
@@ -0,0 +1,176 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/guacsec/guac/pkg/handler/collector/s3/bucket"
+	"github.com/guacsec/guac/pkg/handler/collector/s3/messaging"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+)
+
+const (
+	S3CollectorType = "S3CollectorType"
+)
+
+type S3Collector struct {
+	config S3CollectorConfig
+}
+
+type S3CollectorConfig struct {
+	MessageProvider     string
+	MessageProviderHost string
+	MessageProviderPort string
+	S3Host              string
+	S3Port              string
+	Region              string // optional (defaults to us-east-1, assumes same region for s3 and sqs)
+	Queues              string
+	MpBuilder           messaging.MessageProviderBuilder // optional
+	BucketBuilder       bucket.BuildBucket               // optional
+	SigChan             chan os.Signal                   // optional
+}
+
+func NewS3Collector(cfg S3CollectorConfig) (*S3Collector, error) {
+	s3collector := &S3Collector{
+		config: cfg,
+	}
+	return s3collector, nil
+}
+
+func (s S3Collector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+	queues := strings.Split(s.config.Queues, ",")
+	downloader := getDownloader(s)
+	sigChan := s.config.SigChan
+
+	var wg sync.WaitGroup
+	for _, queue := range queues {
+		wg.Add(1)
+
+		go func(queue string) {
+			mp, err := getMessageProvider(s, queue)
+			if err != nil {
+				logger.Errorf("error getting message provider for queue %v: %v", queue, err)
+				return
+			}
+
+			for {
+				select {
+				case <-sigChan:
+					logger.Infof("Shutting down collector for queue %s...\n", queue)
+					break
```

Comment:
It's good to have this here, but it needs to be setup for `Ctrl-C`. The collector should cleanly shut down then.

Also, Ideally there are two versions of each collector `guacone collect <name>` is meant to be a user run collector that runs once and collects the files in the bucket/source, then exits. The `guaccollect <name>` command is meant to be run as a daemon/service and stay running and watch the bucket/source for new files.

Reply:
@jeffmendoza I wasn't aware that difference between `guacone collect` and `guaccollect` was in the daemon and user use case. In my mind it was difference between direct gql ingest vs nats+ingestor pipeline. For example osv certifier is also used as a running daemon with guacone.
In any case, is it then better to move this command to guaccollect and with current gql ingestion? And then in a subsequent PR implement a `guacone collect s3` command that collects directly from the bucket (without sqs/kafka notifications)?

Reply:
@dejanb You are right, the fundamental difference between the two is direct gql vs nats+ingestor pipeline, and it should stay like that for now.

This expanded into default poll behavior though, assuming the collector that uses nats (guaccollect) should be part of a deployment and run as a service, it should default poll (with optional one-time), and conversely the collector that is direct gql (guacone collect) could be run anywhere, and likely by a user, it should default to run-once (with optional poll). The two cli options are:
https://github.com/guacsec/guac/blob/main/pkg/cli/store.go#L71-L72

In general "guacone" should be the do-all cli for user oriented commands, and service-oriented binaries should be more unixy (do one thing well), and seperate.

Certifiers don't have this yet, but I think they should.

This is covered a bit in these issues: https://github.com/guacsec/guac/issues/719 https://github.com/guacsec/guac/issues/809 I need to work on converting these to developer docs.

Reply:
Thanks @jeffmendoza. So are you OK with leaving things as is in this PR for now? For me it would be ideal that `guaccollect` have options of nats, graphql or ingestor API (#1177) when it's ready. And then we can refactor all collectors to use any of the configured options.
I wouldn't like to refactor this collector to use nats as we are doing this to avoid that exact scenario. So we would like to have option for a long running collector that uses gql. Please let us know what is the best place for this at the moment.

Reply:
Just opened up #1324 to cover some of this.

Yes, we should do just this `guacone` collector first that uses direct gql ingest. No need to also do `guaccollect` at this time, that can be done later if needed.

I do think we need this default to non-polling. Adding the polling option is great, and agree that this should be able to do so. When polling is enabled, make sure to respect `SIGINT/SIGTERM`.

Reply:
@jeffmendoza I have added non-polling behaviour and set it as a default. Non-polling behaviour will download a single file from a specified bucket from S3 and ingest it into guac, then return. I guess in the future this chunk will stay here and the polling logic will be moved to guaccollect.